### PR TITLE
Gen tabular datasets

### LINKED ISSUE #259: Tabular Data Generating Func for Benchmarks?
Would it be helpful for benchmarking to have a tabular data generating function?

For completeness this is what I am thinking can be added to the datasets module:

```
struct DataSpecs {
    rows: usize,
    cols: usize,
    targets: usize,
    feat_distribution: Uniform<f32>,
    target_distribution: Uniform<i8>,
}

impl DataSpecs {
    fn new(
            rows: usize,
            cols: usize,
            targets: usize,
            feat_distribution: Uniform<f32>,
            target_distribution: Uniform<i8>,
        ) -> DataSpecs {
        
            DataSpecs {
                rows,
                cols,
                targets,
                feat_distribution,
                target_distribution
            }
    }
}
```

and something like this be added to the generate.rs

```
fn make_dataset(data_specs: DataSpecs) -> Dataset<f32, i8> {
    let features = Array::random((data_specs.rows, data_specs.cols), data_specs.feat_distribution);
    let targets = Array::random((data_specs.rows, data_specs.targets), data_specs.target_distribution);

    Dataset::new(features, targets)
```

where this would be an example usage

```
fn main() {
    let data_specs = DataSpecs::new(10, 10, 1, Uniform::new(-5., 5.), Uniform::new(0, 2));
    println!("{:#?}", data_specs);

    let data = make_dataset(data_specs);

    println!("{:#?}", data);
}
```

Alternatively, we could have the following instead of a new method implementation to instantiate the struct with keyword arguments. It might be more user friendly if this is a helpful feature for development efforts.

```
let data_specs = DataSpecs{ rows: 10, cols: 10, targets: 1, feat_distribution: Uniform::new(-5., 5.), target_distribution: Uniform::new(0, 2)};
```

### PR DESCRIPTION
**Pull request by oojo12**
### Goal:
This PR aims to make writing benchmarking code a bit easier for developers by parametrizing the datasets dimension.

### Additions to linfa-datasets:

1.  A function to allow generation of fictitious tabular datasets. 
2.  Two optional dependencies (**statrs** and **rand**)

resolves #259 

### Example
```
use ndarray::Array;
use linfa::Dataset;
use ndarray_rand::RandomExt;
use statrs::distribution::{Binomial, DiscreteUniform};
use rand::distributions::Distribution as randDistribution;

fn main() {
    let feat_distr = Binomial::new(0.5, 5 ).unwrap();
    let target_distr = DiscreteUniform::new(0, 5).unwrap();
    let data = make_dataset(5, 5, 2, feat_distr, target_distr);
    
    println!("{:#?}", data);
}

fn make_dataset<X, Y>(num_rows: usize, num_feats: usize, num_targets: usize, feat_distr: X, target_distr: Y) -> Dataset<f64, f64>
where
    X: randDistribution<f64>,
    Y: randDistribution<f64>
{
    let features = Array::random((num_rows, num_feats), feat_distr);
    let targets = Array::random((num_rows, num_targets), target_distr);

    println!("{:#?}", features);
    println!("{:#?}", targets);

    Dataset::new(features, targets)
}
```

Associated output

```
[[1.0, 2.0, 0.0, 3.0, 2.0],
 [3.0, 1.0, 2.0, 4.0, 2.0],
 [2.0, 3.0, 5.0, 2.0, 2.0],
 [2.0, 1.0, 4.0, 2.0, 2.0],
 [2.0, 4.0, 3.0, 1.0, 3.0]], shape=[5, 5], strides=[5, 1], layout=Cc (0x5), const ndim=2
[[3.0, 4.0],
 [5.0, 3.0],
 [5.0, 5.0],
 [2.0, 1.0],
 [2.0, 5.0]], shape=[5, 2], strides=[2, 1], layout=Cc (0x5), const ndim=2
DatasetBase {
    records: [[1.0, 2.0, 0.0, 3.0, 2.0],
     [3.0, 1.0, 2.0, 4.0, 2.0],
     [2.0, 3.0, 5.0, 2.0, 2.0],
     [2.0, 1.0, 4.0, 2.0, 2.0],
     [2.0, 4.0, 3.0, 1.0, 3.0]], shape=[5, 5], strides=[5, 1], layout=Cc (0x5), const ndim=2,
    targets: [[3.0, 4.0],
     [5.0, 3.0],
     [5.0, 5.0],
     [2.0, 1.0],
     [2.0, 5.0]], shape=[5, 2], strides=[2, 1], layout=Cc (0x5), const ndim=2,
    weights: [], shape=[0], strides=[0], layout=CFcf (0xf), const ndim=1,
    feature_names: [],
}
```

the feature array can look like this with a Laplace distribution so we are able to handle both continuous and discrete cases
```
[[-4.360595740165293, 8.155513405816892, -8.478208753936332, 17.23637839788624, 1.7384987675286867],
 [2.6468478875196775, -0.0055614415309673815, 3.6241064859319967, 5.072362710780806, 7.617604811173227],
 [0.79789581354939, -1.4199269716778213, 0.18709110069858814, 1.4115467400849417, -0.8056722910319407],
 [-11.46311409107569, 3.2708729527840843, 4.16040511713887, 1.6879416430935898, 2.451106903259817],
 [-2.7718365458998515, 1.941470709966388, 14.061929511357462, 5.111381148046025, 0.7341567591021234]], shape=[5, 5], strides=[5, 1], layout=Cc (0x5), const ndim=2
```

Adjustments to replicate with the above example

```
use statrs::distribution::{Binomial, DiscreteUniform, Laplace};
let feat_distr = Laplace::new(0.5, 5. ).unwrap();
```

### Outstanding Questions

1. Should an example be added to an "examples" directory?
2.  How do we go about documenting?

### CODE REVIEW

File: datasets/Cargo.toml [hunk 1]
```diff
@@ -14,6 +14,9 @@ ndarray-csv = "=0.5.1"
 csv = "1.1"
 flate2 = "1.0"
 ndarray-rand = { version = "0.14", optional = true }
+statrs = {version = "0.16.0", optional = true }
```
Comment (YuhanLiin) [inline datasets/Cargo.toml:17]: I don't see this being used in the code

File: datasets/Cargo.toml [hunk 2]
```diff
@@ -14,6 +14,9 @@ ndarray-csv = "=0.5.1"
...
+rand = { version = "0.8", optional = true }
```
Comment (YuhanLiin) [inline datasets/Cargo.toml:18]: `rand` is re-exported from `ndarray-rand`, so this dependency isn't needed

**Comment by YuhanLiin** [comment]
For documentation just describe what the function generates. Something like: "This function generates a random, multi-target dataset with the specified number of rows, features, and targets. The values of the features and targets are determined using the provided distributions". You can just put the example into the function docs as a doctest.

File: datasets/Cargo.toml (hunk 1 above)
Comment (oojo12) [reply datasets/Cargo.toml:17]: The distributions are coming from that package. Specifically from this `use statrs::distribution::{Binomial, DiscreteUniform};`

File: datasets/Cargo.toml (hunk 2 above)
Comment (oojo12) [reply datasets/Cargo.toml]: gotcha removed.

**Comment by oojo12** [comment]
cool I'll do that tomorrow.

File: datasets/Cargo.toml (hunk 1 above)
Comment (YuhanLiin) [reply datasets/Cargo.toml:17]: If the user wants those distributions then they can add the dependency in their own code. No need to include it in our crate.

Comment (YuhanLiin) [reply datasets/Cargo.toml:17]: If it's used in the examples, then that's fine

File: datasets/src/generate.rs [hunk 3]
```diff
@@ -53,3 +56,37 @@ fn make_blob(
...
 }
+
+/// Generates a random Linfa::Dataset (ds). The ds values are determined by the provided statistical distributions.
+///
+/// Arguments order:
+///     1 - num_rows
+///     2 - num_feats
+///     3 - num_targets
+///     4 - feat_distr
+///     5 - target_distr
```
Comment (YuhanLiin) [inline datasets/src/generate.rs:67]: Don't need this, since users can just read the parameter list.

**Review by YuhanLiin — APPROVED** [review]

**Pull request merged**